File: internal/library/library_test.go
package library

import (
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"strings"
	"testing"
	"time"
)

// unitMargin is the formula from the design doc, the one a person is most
// likely to have written first.
func unitMargin() Formula {
	return Formula{
		ID:   "unit-margin",
		Name: "Unit margin",
		Kind: KindColumn,
		Expr: "(price - cost) / price",
		Refs: []string{"price", "cost"},
	}
}

func read(t *testing.T, path string) []byte {
	t.Helper()
	b, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read %s: %v", filepath.Base(path), err)
	}
	return b
}

// A formula is a file so that it can be sent to someone, which is only worth
// anything if what comes back out is what went in.
func TestASavedFormulaReadsBackWithEveryFieldIntact(t *testing.T) {
	dir := t.TempDir()
	f := unitMargin()

	if err := Save(dir, f); err != nil {
		t.Fatalf("Save: %v", err)
	}
	got, err := Read(filepath.Join(dir, "unit-margin.unof"))
	if err != nil {
		t.Fatalf("Read: %v", err)
	}

	if got.Created.IsZero() || got.Modified.IsZero() {
		t.Errorf("created = %v, modified = %v, want both stamped", got.Created, got.Modified)
	}
	want := f
	want.Format = formatVersion
	want.Created, want.Modified = got.Created, got.Modified
	if !reflect.DeepEqual(got, want) {
		t.Errorf("read back\n %+v\nwant\n %+v", got, want)
	}
}

// A formula was written once and is edited many times. The first of those facts
// is the one an autosave must not overwrite with the second.
func TestResavingKeepsTheTimeTheFormulaWasFirstWritten(t *testing.T) {
	dir := t.TempDir()
	path := filepath.Join(dir, "unit-margin.unof")

	if err := Save(dir, unitMargin()); err != nil {
		t.Fatalf("Save: %v", err)
	}
	first, err := Read(path)
	if err != nil {
		t.Fatalf("Read: %v", err)
	}

	edited := first
	edited.Expr = "(price - cost) / cost"
	if err := Save(dir, edited); err != nil {
		t.Fatalf("resave: %v", err)
	}
	again, err := Read(path)
	if err != nil {
		t.Fatalf("reread: %v", err)
	}

	if !again.Created.Equal(first.Created) {
		t.Errorf("created = %v, want the first save's %v", again.Created, first.Created)
	}
	if again.Modified.Before(first.Modified) {
		t.Errorf("modified = %v, went backwards from %v", again.Modified, first.Modified)
	}
	if again.Expr != edited.Expr {
		t.Errorf("expr = %q, want the edit %q", again.Expr, edited.Expr)
	}
}

// A notation formula reads no columns, so its file says nothing about columns.
// An empty refs list would still be a claim that it could have some, which is
// why this reads the bytes rather than the decoded value.
func TestANotationFormulaCarriesNoRefsKeyAtAll(t *testing.T) {
	dir := t.TempDir()

	err := Save(dir, Formula{
		ID:   "std-deviation",
		Name: "Std. deviation",
		Kind: KindNotation,
		Expr: `\sigma = \sqrt{\frac{1}{n}\sum_{i=1}^{n}(x_i - \mu)^2}`,
	})
	if err != nil {
		t.Fatalf("Save: %v", err)
	}

	b := read(t, filepath.Join(dir, "std-deviation.unof"))
	if strings.Contains(string(b), "refs") {
		t.Errorf("a notation formula wrote a refs key:\n%s", b)
	}

	var keys map[string]json.RawMessage
	if err := json.Unmarshal(b, &keys); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	if _, ok := keys["refs"]; ok {
		t.Error("refs survived as a key")
	}
}

// An older uno opening a file a newer one wrote must carry what it could not
// read through to the next save. Dropping it would mean this build silently
// deleting a stranger's work every time it autosaved their formula.
func TestKeysThisBuildDoesNotKnowSurviveASaveAndRead(t *testing.T) {
	dir := t.TempDir()
	path := filepath.Join(dir, "unit-margin.unof")

	// Written by hand, as a newer uno would have written it.
	handmade := `{
  "format": 1,
  "id": "unit-margin",
  "name": "Unit margin",
  "kind": "column",
  "expr": "(price - cost) / price",
  "refs": ["price", "cost"],
  "created": "2026-08-31T09:14:02Z",
  "modified": "2026-08-31T16:41:55Z",
  "tolerance": 0.001,
  "provenance": {"by": "a newer uno", "rounds": 3}
}`
	if err := os.WriteFile(path, []byte(handmade), 0o644); err != nil {
		t.Fatalf("seed: %v", err)
	}

	// Read, save, read, save: the second cycle is where a build that only kept
	// unknown keys in memory would lose them.
	for cycle := range 2 {
		f, err := Read(path)
		if err != nil {
			t.Fatalf("cycle %d: Read: %v", cycle, err)
		}
		if err := Save(dir, f); err != nil {
			t.Fatalf("cycle %d: Save: %v", cycle, err)
		}

		var got map[string]any
		if err := json.Unmarshal(read(t, path), &got); err != nil {
			t.Fatalf("cycle %d: unmarshal: %v", cycle, err)
		}
		if got["tolerance"] != 0.001 {
			t.Errorf("cycle %d: tolerance = %v, want 0.001", cycle, got["tolerance"])
		}
		prov, ok := got["provenance"].(map[string]any)
		if !ok {
			t.Fatalf("cycle %d: provenance = %v, want the object it was", cycle, got["provenance"])
		}
		if prov["by"] != "a newer uno" || prov["rounds"] != float64(3) {
			t.Errorf("cycle %d: provenance = %v, want it whole", cycle, prov)
		}
		// The keys it does know are still written once each, not twice.
		if n := strings.Count(string(read(t, path)), `"expr"`); n != 1 {
			t.Errorf("cycle %d: expr appears %d times, want 1", cycle, n)
		}
	}
}

// The autosave writes while someone is typing, which is when a crash is most
// likely. An interrupted one must cost the keystroke, never the formula that
// was already saved — the same promise safefile makes to the .uno container.
func TestAFailedSaveLeavesThePreviousFormulaWhole(t *testing.T) {
	dir := t.TempDir()
	path := filepath.Join(dir, "unit-margin.unof")

	if err := Save(dir, unitMargin()); err != nil {
		t.Fatalf("Save: %v", err)
	}
	saved := read(t, path)

	// A formula that cannot be encoded at all: the failure lands mid-save, with
	// the previous file sitting on disk beside it.
	broken := unitMargin()
	broken.Expr = "the new one"
	broken.Extra = map[string]json.RawMessage{"nonsense": json.RawMessage("not json")}
	if err := Save(dir, broken); err == nil {
		t.Fatal("want a failure, got nil")
	}

	if got := read(t, path); string(got) != string(saved) {
		t.Errorf("the previous formula changed:\n%s\nwant\n%s", got, saved)
	}
	left, err := os.ReadDir(dir)
	if err != nil {
		t.Fatalf("read dir: %v", err)
	}
	if len(left) != 1 || left[0].Name() != "unit-margin.unof" {
		t.Errorf("directory holds %v, want only the formula", left)
	}
}

// The id becomes a filename, so it is the one field in a .unof that can reach
// outside the library. Formulas arrive from other people, so this is a trust
// boundary and not a formality.
func TestAnIDThatCouldNameAPathIsRefused(t *testing.T) {
	for _, id := range []string{
		"",
		"../escape",
		"a/b",
		`a\b`,
		".",
		"..",
		"...",
		".hidden",
		"c:margin",
		"unit\nmargin",
		strings.Repeat("x", maxIDLen+1),
	} {
		dir := t.TempDir()
		f := unitMargin()
		f.ID = id

		if err := Save(dir, f); err == nil {
			t.Errorf("id %q was accepted", id)
		}

		// Nothing was written anywhere, including beside the directory rather
		// than in it, which is what an escaping id would have managed.
		for _, look := range []string{dir, filepath.Dir(dir)} {
			entries, err := os.ReadDir(look)
			if err != nil {
				t.Fatalf("read dir: %v", err)
			}
			for _, e := range entries {
				if strings.HasSuffix(e.Name(), ext) {
					t.Errorf("id %q wrote %s", id, filepath.Join(look, e.Name()))
				}
			}
		}
	}
}

// A .unof holds no path for the reason a .uno holds none: a file that only
// works where it was written is not reusable anywhere, and the whole point of
// one file per formula is that the file can be sent to someone else.
func TestAFormulaCarriesNoPathFromTheMachineThatWroteIt(t *testing.T) {
	dir := t.TempDir()

	if err := Save(dir, unitMargin()); err != nil {
		t.Fatalf("Save: %v", err)
	}

	b := string(read(t, filepath.Join(dir, "unit-margin.unof")))
	for _, path := range []string{dir, filepath.Dir(dir), os.TempDir()} {
		if strings.Contains(b, path) {
			t.Errorf("the file names %s:\n%s", path, b)
		}
	}
	// No value in the file begins with a separator either. A division reads as
	// "a / b" and never as "\"/", so this catches an absolute path without
	// catching the arithmetic.
	if strings.Contains(b, `"`+string(filepath.Separator)) {
		t.Errorf("the file holds an absolute path:\n%s", b)
	}
}

// One file per formula is worth having only if one bad file is one formula. A
// library that refused to open because a single .unof was truncated would have
// given away the thing that made the layout worth choosing.
func TestOneUnreadableFileCostsOneFormulaAndNotTheLibrary(t *testing.T) {
	dir := t.TempDir()

	for _, id := range []string{"unit-margin", "strip-thousands"} {
		f := unitMargin()
		f.ID, f.Name = id, id
		if err := Save(dir, f); err != nil {
			t.Fatalf("Save %s: %v", id, err)
		}
	}
	if err := os.WriteFile(filepath.Join(dir, "truncated.unof"), []byte(`{"id": "trunc`), 0o644); err != nil {
		t.Fatalf("seed: %v", err)
	}
	// Not a formula, and not this package's business either way.
	if err := os.WriteFile(filepath.Join(dir, "notes.txt"), []byte("hello"), 0o644); err != nil {
		t.Fatalf("seed: %v", err)
	}

	got, err := Load(dir)
	if err == nil {
		t.Error("the truncated file was not reported")
	} else if !strings.Contains(err.Error(), "truncated.unof") {
		t.Errorf("error = %v, want it to name the file", err)
	}

	var ids []string
	for _, f := range got {
		ids = append(ids, f.ID)
	}
	// Sorted, because the order a directory hands back is not an order.
	if want := []string{"strip-thousands", "unit-margin"}; !reflect.DeepEqual(ids, want) {
		t.Errorf("loaded %v, want %v", ids, want)
	}
}

// A person who has never written a formula has no folder, which is not a fault
// to report to them.
func TestALibraryThatDoesNotExistYetIsEmptyRatherThanBroken(t *testing.T) {
	got, err := Load(filepath.Join(t.TempDir(), "formulas"))
	if err != nil {
		t.Errorf("Load: %v", err)
	}
	if len(got) != 0 {
		t.Errorf("loaded %v, want nothing", got)
	}
}

// A build that guessed at a layout it does not know would either misread the
// file or save what it misread back over it, and the person would be told
// neither.
func TestAFormulaFromANewerUnoIsRefusedByName(t *testing.T) {
	dir := t.TempDir()
	path := filepath.Join(dir, "unit-margin.unof")

	newer := `{"format": 99, "id": "unit-margin", "name": "Unit margin", "kind": "column", "expr": "1"}`
	if err := os.WriteFile(path, []byte(newer), 0o644); err != nil {
		t.Fatalf("seed: %v", err)
	}

	_, err := Read(path)
	if err == nil {
		t.Fatal("want a refusal, got nil")
	}
	if !strings.Contains(err.Error(), "unit-margin.unof") || !strings.Contains(err.Error(), "99") {
		t.Errorf("error = %v, want it to name the file and the format", err)
	}
}

// A file whose id names a path is a file this package must not hand out, so the
// check happens on the way in as well as on the way out.
func TestAFileWhoseIDNamesAPathIsRefusedOnRead(t *testing.T) {
	dir := t.TempDir()
	path := filepath.Join(dir, "escape.unof")

	if err := os.WriteFile(path, []byte(`{"format": 1, "id": "../escape", "kind": "column"}`), 0o644); err != nil {
		t.Fatalf("seed: %v", err)
	}
	if _, err := Read(path); err == nil {
		t.Fatal("want a refusal, got nil")
	}
}

// The times are written the way the .uno manifest writes them: UTC, whole
// seconds, so a diff between two saves shows what changed rather than a
// nanosecond nobody typed.
func TestTheTimesAreWrittenAsWholeSecondsInUTC(t *testing.T) {
	dir := t.TempDir()

	if err := Save(dir, unitMargin()); err != nil {
		t.Fatalf("Save: %v", err)
	}

	var got struct {
		Created  string `json:"created"`
		Modified string `json:"modified"`
	}
	if err := json.Unmarshal(read(t, filepath.Join(dir, "unit-margin.unof")), &got); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	for _, stamp := range []string{got.Created, got.Modified} {
		if _, err := time.Parse("2006-01-02T15:04:05Z", stamp); err != nil {
			t.Errorf("stamp %q is not a whole second in UTC: %v", stamp, err)
		}
	}
}

// Markdown was chosen so a .unof stays legible to someone reading it in a diff
// or a chat window without uno. A comparison rewritten as < would take
// that back one character at a time.
func TestAComparisonIsWrittenAsItWasTyped(t *testing.T) {
	dir := t.TempDir()
	f := unitMargin()
	f.Expr = "price < cost & margin > 0"

	if err := Save(dir, f); err != nil {
		t.Fatalf("Save: %v", err)
	}

	b := string(read(t, filepath.Join(dir, "unit-margin.unof")))
	if !strings.Contains(b, f.Expr) {
		t.Errorf("the expression was rewritten:\n%s", b)
	}
	got, err := Read(filepath.Join(dir, "unit-margin.unof"))
	if err != nil {
		t.Fatalf("Read: %v", err)
	}
	if got.Expr != f.Expr {
		t.Errorf("expr = %q, want %q", got.Expr, f.Expr)
	}
}

File: internal/grid/tests/library/library.test.ts
import { describe, expect, test } from "vite-plus/test";

import { FORMAT_VERSION, formatFormula, parseFormula, validID } from "../../src/library/index.ts";
import type { Formula } from "../../src/library/index.ts";

function column(id: string, name: string, expr: string, refs?: string[]): Formula {
  return {
    format: 0,
    id,
    name,
    kind: "column",
    expr,
    refs,
    created: undefined,
    modified: undefined,
  };
}

/** What a save-then-read round trip does, without a filesystem in the way. */
function roundTrip(f: Formula): Formula {
  const { text } = formatFormula(f);
  return parseFormula(f.id + ".unof", text);
}

test("a saved formula reads back with every field intact", () => {
  const f = column("unit-margin", "Unit margin", "(price - cost) / price", ["cost", "price"]);
  const back = roundTrip(f);

  expect(back.id).toBe("unit-margin");
  expect(back.name).toBe("Unit margin");
  expect(back.kind).toBe("column");
  expect(back.expr).toBe("(price - cost) / price");
  expect(back.refs).toEqual(["cost", "price"]);
  expect(back.format).toBe(FORMAT_VERSION);
  expect(back.created).toBeDefined();
  expect(back.modified).toBeDefined();
});

// Modified is what this save is; created is filled in only the first time, so a
// formula cannot come to claim it was written after it was last edited.
test("resaving keeps the time the formula was first written", () => {
  const first = formatFormula(column("f", "F", "a + b")).stamped;
  expect(first.created).toBeDefined();

  const second = formatFormula(first).stamped;
  expect(second.created).toEqual(first.created);
});

// A notation formula reads nothing, so it carries no refs key at all rather
// than an empty list that would imply it could.
test("a notation formula carries no refs key at all", () => {
  const f: Formula = {
    format: 0,
    id: "variance",
    name: "Variance",
    kind: "notation",
    expr: "\\sigma^2",
    created: undefined,
    modified: undefined,
  };

  const { text } = formatFormula(f);
  expect(text).not.toContain("refs");
  expect(roundTrip(f).refs).toBeUndefined();
});

// An older uno opening a file written by a newer one must not quietly drop what
// it could not read and then write that loss back over the file.
test("keys this build does not know survive a save and read", () => {
  const text = JSON.stringify(
    {
      format: 1,
      id: "f",
      name: "F",
      kind: "column",
      expr: "a + b",
      created: "2026-01-01T00:00:00Z",
      modified: "2026-01-01T00:00:00Z",
      colour: "blue",
      futureThing: { nested: true },
    },
    undefined,
    2,
  );

  const f = parseFormula("f.unof", text);
  expect(f.extra?.get("colour")).toBe("blue");

  const written = JSON.parse(formatFormula(f).text) as Record<string, unknown>;
  expect(written["colour"]).toBe("blue");
  expect(written["futureThing"]).toEqual({ nested: true });
});

// Unknown keys are written in name order, so a save that changed nothing
// produces the same bytes as the one before it.
test("unknown keys are written in a stable order", () => {
  const text = JSON.stringify({
    format: 1,
    id: "f",
    name: "F",
    kind: "column",
    expr: "a",
    zeta: 1,
    alpha: 2,
    mu: 3,
  });

  const f = parseFormula("f.unof", text);
  const written = formatFormula(f).text;
  expect(written.indexOf('"alpha"')).toBeLessThan(written.indexOf('"mu"'));
  expect(written.indexOf('"mu"')).toBeLessThan(written.indexOf('"zeta"'));
});

// Formulas arrive from other people -- that is the point of making each one a
// file -- so the id is checked before it is ever joined to a path.
describe("an id that could name a path is refused", () => {
  const bad = [
    "",
    ".",
    "..",
    ".hidden",
    "../escape",
    "a/b",
    "a\\b",
    "C:name",
    "with\u0000null",
    "x".repeat(201),
  ];

  for (const id of bad) {
    test(JSON.stringify(id), () => {
      expect(() => validID(id)).toThrow();
    });
  }

  test("an ordinary id is accepted", () => {
    expect(() => validID("unit-margin")).not.toThrow();
    expect(() => validID("marge unitaire")).not.toThrow();
  });
});

// The id is checked on the way in as well as on the way out, so no id that
// could name a path is ever handed to a caller in the first place.
test("a file whose id names a path is refused on read", () => {
  const text = JSON.stringify({ format: 1, id: "../escape", name: "F", kind: "column", expr: "a" });
  expect(() => parseFormula("f.unof", text)).toThrow();
});

test("a formula from a newer uno is refused by name", () => {
  const text = JSON.stringify({ format: FORMAT_VERSION + 1, id: "f", name: "F", kind: "column" });

  let thrown: Error | undefined;
  try {
    parseFormula("f.unof", text);
  } catch (err) {
    thrown = err as Error;
  }
  expect(thrown).toBeDefined();
  expect(thrown!.message).toContain("f.unof");
  expect(thrown!.message).toContain("newer uno");
});

// A file that only works where it was written is not reusable anywhere.
test("a formula carries no path from the machine that wrote it", () => {
  const { text } = formatFormula(column("f", "F", "a + b", ["a", "b"]));
  expect(text).not.toContain("/home/");
  expect(text).not.toContain("path");
  expect(text).not.toContain("dir");
});

test("the times are written as whole seconds in UTC", () => {
  const { text } = formatFormula(column("f", "F", "a"));
  const written = JSON.parse(text) as Record<string, string>;

  expect(written["created"]).toMatch(/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$/);
  expect(written["modified"]).toMatch(/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$/);
});

// Escaping a "<" would cost exactly the legibility that made markdown the right
// thing to store.
test("a comparison is written as it was typed", () => {
  const { text } = formatFormula(column("f", "F", "a < b"));
  expect(text).toContain("a < b");
  expect(text).not.toContain("\\u003c");
});

test("a file that is not readable is refused by name", () => {
  expect(() => parseFormula("broken.unof", "{not json")).toThrow(/broken\.unof/);
  expect(() => parseFormula("broken.unof", "[1,2,3]")).toThrow(/broken\.unof/);
});
